src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/EnumShouldNotHaveDuplicatedValues.cs

```diff
@@ -0,0 +1,150 @@
+                        !fieldInitializerOperation.Value.ConstantValue.HasValue)
+                    {
+                        return;
+                    }
+
+                    var onlyReferencesOneField = GetFilteredDescendants(fieldInitializerOperation, op => op.Kind != OperationKind.Binary)
+                        .OfType<IFieldReferenceOperation>()
+                        .Any();
+                    if (onlyReferencesOneField)
+                    {
+                        return;
+                    }
+
+                    invalidMembersByValue.AddOrUpdate(fieldInitializerOperation.Value.ConstantValue.Value,
+                        new List<SyntaxNode> { fieldInitializerOperation.Syntax.Parent },
+                        (key, value) =>
+                        {
+                            value.Add(fieldInitializerOperation.Syntax.Parent);
+                            return value;
+                        });
+                }, OperationKind.FieldInitializer);
+
+                // ...and report at the end of the enum declaration
+                ssac.RegisterSymbolEndAction(sac =>
+                {
+                    foreach (var kvp in invalidMembersByValue)
+                    {
+                        foreach (var item in kvp.Value.OrderBy(x => x.GetLocation().SourceSpan).Skip(1))
+                        {
+                            sac.ReportDiagnostic(item.CreateDiagnostic(RuleDuplicatedValue));
+                        }
+                    }
+                });
+
+                // Collect and report on duplicated bitwise parts
+                ssac.RegisterOperationAction(oc =>
+                {
+                    var binaryOperation = (IBinaryOperation)oc.Operation;
+                    if (binaryOperation.OperatorKind != BinaryOperatorKind.Or)
+                    {
+                        return;
+                    }
+
+                    var bitwiseRefFieldsByField = GetFilteredDescendants(binaryOperation,
+                            op => op.Kind != OperationKind.Binary || (op is IBinaryOperation binaryOp && binaryOp.OperatorKind == BinaryOperatorKind.Or))
+                        .OfType<IFieldReferenceOperation>()
+                        .GroupBy(fro => fro.Field, fro => fro.Syntax)
+                        .ToDictionary(x => x.Key, x => x.ToList());
+
+                    foreach (var kvp in bitwiseRefFieldsByField)
+                    {
+                        foreach (var item in kvp.Value.OrderBy(x => x.GetLocation().SourceSpan).Skip(1))
+                        {
+                            oc.ReportDiagnostic(item.CreateDiagnostic(RuleDuplicatedBitwiseValuePart));
+                        }
+                    }
+                }, OperationKind.Binary);
+            }, SymbolKind.NamedType);
+        }
+
+        private static IEnumerable<IOperation> GetFilteredDescendants(IOperation operation, Func<IOperation, bool> descendIntoOperation)
```

Comment:
This is a copy of the `Descendant` method where I have replaced `Push`/`Pop` calls with `Add`/`Get`/`Remove` calls as the `ArrayBuilder` is different from the one in Roslyn.

Is it worth suggesting this predicate overload on Roslyn?

Reply:
ping @mavasani

Reply:
I think we don't want to encourage people to use the Children or Descendants API unless they really need to. You can possibly open an issue for it in Roslyn but it may not get traction.